src/EFCore.PG/Scaffolding/Internal/NpgsqlDatabaseModelFactory.cs

```diff
@@ -543,9 +544,32 @@ ELSE pg_get_expr(indpred, cls.oid)
                             .GetValueOrDefault<uint[]>("indclass")
                             .Select(oid => opClasses.TryGetValue(oid, out var opc) && !opc.IsDefault ? opc.Name : null)
                             .ToArray();
+
                         if (opClassNames.Any(op => op != null))
                             index[NpgsqlAnnotationNames.IndexOperators] = opClassNames;
 
+                        var options = record.GetValueOrDefault<ushort[]>("indoption");
+
+                        // The first bit specifies whether values are sorted in descending order.
```

Comment:
[According to the PostgreSQL docs](https://www.postgresql.org/docs/current/catalog-pg-index.html), the "meaning of the bits is defined by the index's access method." - is this code specific to btree? If so we may need to switch on the different methods and this becomes a bit more complicated.

Maybe as a first step, we can just parameterize the scaffolding tests for the different methods (i.e. run the exact same scaffolding tests on btree, gist...). It may very well be that all methods share these two specific common bits.

Reply:
> is this code specific to btree?

No, it's not. Should we restrict it to `btree` to err on the safe side?

> It may very well be that all methods share these two specific common bits.

[pg_index.h](https://doxygen.postgresql.org/pg__index_8h_source.html) says:

> ```
> /*
>  * Index AMs that support ordered scans must support these two indoption
>  * bits.  Otherwise, the content of the per-column indoption fields is
>  * open for future definition.
>  */
> ```

Now, I must admit I have no clue which index methods that do or don't support ordered scans.

Reply:
I guess we should check `pg_am.amcanorder`?

Reply:
I really don't know much about this, certainly not more than you :) How about parameterizing the tests for the major index method and seeing what happens? Maybe it'll just work out of the box...

Reply:
I tried parameterizing the tests, but ended up failing for everything (built-in) except `btree`. Tried `gist`, `gin`, `hash` and `brin`.

I guess it makes sense given this:

```sql
SELECT amname, pg_indexam_has_property(oid, 'can_order') AS amcanorder FROM pg_am;
```

|amname|amhandler|amtype|amcanorder|
|------|---------|------|----------|
|btree|bthandler|i|true|
|hash|hashhandler|i|false|
|gist|gisthandler|i|false|
|gin|ginhandler|i|false|
|spgist|spghandler|i|false|
|brin|brinhandler|i|false|

So I guess we could get away with restricting it to `btree`, at least for the built-in access methods, but since it's possible to query `pg_indexam_has_property`/`pg_am.amcanorder`, I think we should support it.

Reply:
Ah, OK - thanks for testing that. Does that mean that ASC/DESC are ignored for non-btree index methods, or does it generate an error?

Otherwise yeah, ideally in scaffolding we'd perform an early loading of supported index methods with their `amcanorder`, and consult that when scaffolding each index.